cg/services/file_delivery/file_formatter_service/utils/sample_file_concatenation_formatter.py

```diff
@@ -0,0 +1,111 @@
+    SampleFileFormatter,
+)
+
+
+class SampleFileConcatenationFormatter(SampleFileFormatter):
+    """
+    Format the sample files to deliver, concatenate fastq files and return the formatted files.
+    Used for workflows: Microsalt and Mutant.
+    """
+
+    def __init__(self, concatenation_service: FastqConcatenationService):
+        self.concatenation_service = concatenation_service
+
+    def format_files(
+        self, moved_files: list[SampleFile], ticket_dir_path: Path
+    ) -> list[FormattedFile]:
+        """Format the sample files to deliver, concatenate fastq files and return the formatted files."""
+        formatted_files: list[FormattedFile] = super().format_files(
+            moved_files=moved_files, ticket_dir_path=ticket_dir_path
+        )
+        forward_paths, reverse_path = self._concatenate_fastq_files(formatted_files=formatted_files)
+        self._replace_fastq_paths(
+            reverse_paths=reverse_path,
+            forward_paths=forward_paths,
+            formatted_files=formatted_files,
+        )
+        return formatted_files
+
+    def _concatenate_fastq_files(
+        self, formatted_files: list[FormattedFile]
+    ) -> tuple[list[Path], list[Path]]:
+        unique_sample_dir_paths: set[Path] = self._get_unique_sample_paths(
+            sample_files=formatted_files
+        )
+        forward_paths: list[Path] = []
+        reverse_paths: list[Path] = []
+        for sample_dir_path in unique_sample_dir_paths:
+            fastq_directory: Path = sample_dir_path
+            sample_name: str = fastq_directory.name
+
+            forward_path: Path = generate_concatenated_fastq_delivery_path(
+                fastq_directory=fastq_directory,
+                sample_name=sample_name,
+                direction=ReadDirection.FORWARD,
+            )
+            forward_paths.append(forward_path)
+            reverse_path: Path = generate_concatenated_fastq_delivery_path(
+                fastq_directory=fastq_directory,
+                sample_name=sample_name,
+                direction=ReadDirection.REVERSE,
+            )
+            reverse_paths.append(reverse_path)
+            self.concatenation_service.concatenate(
+                fastq_directory=fastq_directory,
+                forward_output_path=forward_path,
+                reverse_output_path=reverse_path,
+                remove_raw=True,
+            )
+        return forward_paths, reverse_paths
+
+    @staticmethod
+    def _get_unique_sample_paths(sample_files: list[FormattedFile]) -> set[Path]:
+        sample_paths: list[Path] = []
+        for sample_file in sample_files:
+            sample_paths.append(sample_file.formatted_path.parent)
+        return set(sample_paths)
+
+    @staticmethod
+    def _replace_fastq_formatted_file_path(
+        formatted_files: list[FormattedFile],
+        direction: ReadDirection,
+        new_path: Path,
+    ) -> None:
+        """Replace the formatted file path with the new path."""
```

Comment:
We don't use the original file name during the concatenation but rather just the sample folder. I think it would make it a bit more complicated.

What is the benefit of doing it earlier and the draw back of the current solution?

Reply:
Correct me if I'm wrong but this is the current logic. 

Find files to concatenate -> concatenate files -> find files that have been concatenated -> replace their paths.

The `_replace_fastq_formatted_file_path` seems unnecessary to me since we've already found all the files previously. 

Looking at the code though I can't find a cleaner way to do it. You can resolve if you want

Reply:
The super call moves all sample files into a sample folder in the ticket folder
the concatenation takes the sample folders as input, not the single fastq files, so we don't touch them here
then the new fastq files get created and the old ones deleted and their paths not really touched inside this formatter
---- Untill this point the concatenated fastq files have not existed
now we replace paths